python-frontend/src/main/java/org/sonar/python/semantic/v2/types/AstBasedPropagation.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic.v2.types;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.semantic.v2.SymbolV2;
+import org.sonar.python.semantic.v2.SymbolV2Utils;
+import org.sonar.python.semantic.v2.TypeTable;
+import org.sonar.python.semantic.v2.UsageV2;
+import org.sonar.python.tree.NameImpl;
+import org.sonar.python.tree.TreeUtils;
+import org.sonar.python.types.HasTypeDependencies;
+import org.sonar.python.types.v2.PythonType;
+import org.sonar.python.types.v2.UnionType;
+
+public class AstBasedPropagation {
```

Comment:
This name confuses me cause we also have the `Propagation` abstract class in the same package, but the `AstBasedPropagation` is not a part of the `Propagation` hierarchy.  Probably `AstBasedPropagationProcessor` could be better

Reply:
Or since it is an alternative to the `FlowSensetiveTypeInference` - could be named somehow like `AstBasedProbagationTypeInference`

Reply:
I've renamed it to `AstBasedTypeInference`. The similarity to `FlowSensitiveTypeInference` is a good point

Reply:
I've renamed it to `AstBasedTypeInference`. The similarity to `FlowSensitiveTypeInference` is a good point